content/200-concepts/100-components/02-prisma-client/052-advanced-type-safety/index.mdx

```diff
@@ -0,0 +1,116 @@
+
+```ts
+import { Prisma } from '@prisma/client'
+
+const userEmail: Prisma.UserSelect = {
+  email: true,
+}
+```
+
+## What are generated types?
+
+Generated types are TypeScript types that are derived from your models. They allow you to perform operations such as a `select` or `include` on your records in a typed manner.
+
+`UserSelect` is one such generated type. Its object properties match those that are supported by `select` statements according to the model.
+
+The first tab below shows the `UserSelect` generated type and how each property on the object has a type annotation. The second tab shows the resulting schema model.
+
+<TabbedContent tabs={[<FileWithIcon text="Generated type" icon="code"/>, <FileWithIcon text="Model" icon="prisma"/>]}>
+
+<tab>
+
+```ts
+type Prisma.UserSelect = {
+    id?: boolean | undefined;
+    email?: boolean | undefined;
+    name?: boolean | undefined;
+    posts?: boolean | Prisma.PostFindManyArgs | undefined;
+    profile?: boolean | Prisma.ProfileArgs | undefined;
+}
+```
+
+</tab>
+
+<tab>
+
+```prisma
+model User {
+  id      Int      @id @default(autoincrement())
+  email   String   @unique
+  name    String?
+  posts   Post[]
+  profile Profile?
+}
+```
+
+</tab>
+
+</TabbedContent>
+
+In TypeScript the concept of [type annotations](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-annotations-on-variables) is when you declare a variable and add a type annotation to describe the type of the variable. See the below example.
+
+```ts
+const myAge: number = 37
+const myName: string = 'Rich'
+```
+
+Both of these variable declarations have been given a type annotation to specify what primitive type they are, `number` and `string` respectively. Most of the time this kind of annotation is not needed as TypeScript will infer the type of the variable based on how its initialized. In the above example `myAge` was initialized with a number so TypeScript guesses that it should be typed as a number.
+
+Going back to the `UserSelect` type, if you were to use dot notation on the created object `userEmail`, you would have access to all of the fields on the `User` model that can be interacted with using a `select` statement.
+
+```prisma file=schema.prisma
+model User {
+  id      Int      @id @default(autoincrement())
+  email   String   @unique
+  name    String?
+  posts   Post[]
+  profile Profile?
+}
+```
+
+```ts
+import { Prisma } from '@prisma/client'
+
+const userEmail: Prisma.UserSelect = {
+  email: true,
+}
+
+// properties available on the typed object
+userEmail.id
+userEmail.email
+userEmail.name
+userEmail.posts
+userEmail.profile
+```
+
+In the same mould, you can type an object with an `include` generated type then your object would have access to those properties on which you can use an `include` statement.
+
+```ts
+import { Prisma } from '@prisma/client'
+
+const userInclude: Prisma.UserInclude = {
```

Comment:
sorry to be so picky again, but if we call the previous one `userEmail`, should we call this one `userPosts` for consistency?

Reply:
Yes! totally right!